### PR TITLE
Avoid decrementing backends_count and frontends_count when the value is zero

### LINKED ISSUE #539: Got a panick when the last backend of an application is remove
When we remove the last backend of an application with `sozuctl` I get a panick (see below).
This is easy to reproduce, you just have to create an application with one backend and use `sozuctl backend remove` to reproduce this error.

``` bash
2018-10-11T10:34:52Z 138668649469416 22156 MASTER INFO	Starting up
2018-10-11T10:34:52Z 138668650672640 22156 MASTER INFO	0 launching worker
2018-10-11T10:34:52Z 138668650791122 22156 MASTER INFO	0 worker launched: 22157
2018-10-11T10:34:52Z 138668651115767 22156 MASTER INFO	sending listeners to new worker: Listeners { http: [], tls: [], tcp: [] }
2018-10-11T10:34:52Z 138668651181395 22156 MASTER INFO	sent listeners from master: Ok(())
2018-10-11T10:34:52Z 138668652155800 22156 MASTER INFO	1 launching worker
2018-10-11T10:34:52Z 138668652278516 22156 MASTER INFO	1 worker launched: 22158
2018-10-11T10:34:52Z 138668652513077 22156 MASTER INFO	sending listeners to new worker: Listeners { http: [], tls: [], tcp: [] }
2018-10-11T10:34:52Z 138668652543633 22156 MASTER INFO	sent listeners from master: Ok(())
2018-10-11T10:34:52Z 138668652572171 22156 MASTER INFO	created workers: [Worker { id: 0, run_state: Running }, Worker { id: 1, run_state: Running }]
2018-10-11T10:34:52Z 138668652608451 22156 MASTER ERROR	could not get saved state path: No such file or directory (os error 2)
2018-10-11T10:34:52Z 138668652751825 22156 MASTER ERROR	could not register metrics socket
2018-10-11T10:34:52Z 138668653012273 22156 MASTER INFO	waiting for configuration client connections
2018-10-11T10:34:52Z 138668654158739 22157 WRK-00 INFO	worker 0 starting...
2018-10-11T10:34:52Z 138668655106000 22158 WRK-01 INFO	worker 1 starting...
2018-10-11T10:34:52Z 138668925501972 22158 WRK-01 INFO	will try to receive listeners
2018-10-11T10:34:52Z 138668925653829 22158 WRK-01 INFO	received listeners: Some(Listeners { http: [], tls: [], tcp: [] })
2018-10-11T10:34:52Z 138668925701962 22158 WRK-01 INFO	starting event loop
2018-10-11T10:34:52Z 138668927699445 22157 WRK-00 INFO	will try to receive listeners
2018-10-11T10:34:52Z 138668927838073 22157 WRK-00 INFO	received listeners: Some(Listeners { http: [], tls: [], tcp: [] })
2018-10-11T10:34:52Z 138668927873932 22157 WRK-00 INFO	starting event loop
thread 'main' panicked at 'attempt to subtract with overflow', bin/src/command/orders.rs:663:45
note: Run with `RUST_BACKTRACE=1` for a backtrace.
2018-10-11T10:36:14Z 138750731833165 22158 WRK-01 ERROR	command channel was closed
2018-10-11T10:36:14Z 138750731839671 22157 WRK-00 ERROR	command channel was closed
```
Here is the stack backtrace:

``` bash
thread 'main' panicked at 'attempt to subtract with overflow', bin/src/command/orders.rs:663:45
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: sozu::register_panic_hook::{{closure}}
             at bin/src/main.rs:243
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:481
   6: std::panicking::continue_panic_fmt
             at libstd/panicking.rs:391
   7: rust_begin_unwind
             at libstd/panicking.rs:326
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:77
   9: core::panicking::panic
             at libcore/panicking.rs:52
  10: sozu::command::orders::<impl sozu::command::CommandServer>::worker_order
             at bin/src/command/orders.rs:663
  11: sozu::command::orders::<impl sozu::command::CommandServer>::handle_client_message
             at bin/src/command/orders.rs:66
  12: sozu::command::CommandServer::handle_client_events
             at bin/src/command/mod.rs:514
  13: sozu::command::CommandServer::ready
             at bin/src/command/mod.rs:332
  14: sozu::command::CommandServer::run
             at bin/src/command/mod.rs:226
  15: sozu::command::start
             at bin/src/command/mod.rs:645
  16: sozu::main::{{closure}}
             at bin/src/main.rs:84
  17: <core::result::Result<T, E>>::map
             at /checkout/src/libcore/result.rs:468
  18: sozu::main
             at bin/src/main.rs:63
  19: std::rt::lang_start::{{closure}}
             at /checkout/src/libstd/rt.rs:74
  20: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  21: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:103
  22: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:392
             at libstd/rt.rs:58
  23: std::rt::lang_start
             at /checkout/src/libstd/rt.rs:74
  24: main
  25: __libc_start_main
  26: _start
```

### PR DESCRIPTION
**Pull request by NotBad4U**
When we remove the last backend or frontend with `sozuctl` e.g.:
`./target/debug/sozuctl -c bin/config.toml backend remove --address <Ipaddress> --id <appId> --backend-id <backend-id>`
The counter `backends|frontends` is decremented and that can panic when the value is zero because the counter type is `usize`.

This PR add a guard to avoid decrementing the counter when the value is zero.
This will fix #539 .

### CODE REVIEW

**Comment by Geal** [comment]
this should instead check if a frontend or backend was actually removed.

File: bin/src/command/orders.rs [hunk 1]
```diff
@@ -599,6 +599,25 @@ impl CommandServer {
       ::std::env::set_var("RUST_LOG", logging_filter);
     }
 
+    // Check if the backend or frontend exist before deleting it
+    match order {
```
Comment (Geal) [inline bin/src/command/orders.rs]: if `worker_id` is not `None`, we should not return from here: the message is specifically addressed to one worker. It is useful if an external tool detects that the master process and the worker do not have the same configuration.

File: bin/src/command/orders.rs [hunk 2]
```diff
@@ -658,15 +677,24 @@ impl CommandServer {
...
     match order {
-      ProxyRequestData::AddBackend(_) => self.backends_count += 1,
-      ProxyRequestData::RemoveBackend(_) => self.backends_count -= 1,
-      ProxyRequestData::AddHttpFront(_) | ProxyRequestData::AddHttpsFront(_) | ProxyRequestData::AddTcpFront(_) => {
-        self.frontends_count += 1;
-      },
-      ProxyRequestData::RemoveHttpFront(_) | ProxyRequestData::RemoveHttpsFront(_) | ProxyRequestData::RemoveTcpFront(_) => {
-        self.frontends_count -= 1;
-      },
+      ProxyRequestData::RemoveBackend(ref mut backend) => self.config.applications.get_mut(&backend.app_id).map(|app| app.remove_backend(&backend.address)),
```
Comment (Geal) [inline bin/src/command/orders.rs]: why modify `self.config`? This is an object created from the configuration file, the real configuration state is in `self.state`

File: command/src/config.rs [hunk 3]
```diff
@@ -523,6 +523,26 @@ pub struct HttpAppConfig {
 }
 
 impl HttpAppConfig {
```
Comment (Geal) [inline command/src/config.rs]: the modifications to `command/src/config.rs` are not needed. The `Config` object should not be modified after it has been created

File: bin/src/command/orders.rs [hunk 4]
```diff
@@ -659,14 +680,14 @@ impl CommandServer {
...
-      },
+      ProxyRequestData::AddBackend(_)
+      | ProxyRequestData::RemoveBackend(_) => self.backends_count = self.state.count_backends(),
+      ProxyRequestData::AddHttpFront(_)
+      | ProxyRequestData::AddHttpsFront(_)
+      | ProxyRequestData::AddTcpFront(_)
+      | ProxyRequestData::RemoveHttpFront(_)
+      | ProxyRequestData::RemoveHttpsFront(_)
+      | ProxyRequestData::RemoveTcpFront(_) => self.frontends_count = self.state.count_frontends(),
       _ => {}
```
Comment (Geal) [inline bin/src/command/orders.rs]: why not update the applications count as well?

File: command/src/state.rs [hunk 5]
```diff
@@ -63,6 +63,22 @@ impl ConfigState {
     }
   }
 
+  pub fn contains_backend(&self, app_id: &str, address: &SocketAddr) -> bool {
```
Comment (Geal) [inline command/src/state.rs]: those modifications to `ConfigState` are not needed

Comment (NotBad4U) [reply command/src/state.rs]: Why ?

Comment (Geal) [reply command/src/state.rs]: because they are not used at all in this pull request

Comment (NotBad4U) [reply command/src/state.rs]: Oh yes ! now I don't need it anymore ... good catch :smile:

File: bin/src/command/orders.rs (hunk 4 above)
Comment (Geal) [reply bin/src/command/orders.rs]: please update the applications count as well

**Pull request merged**